common/changeset/account_changeset_utils.go

```diff
@@ -77,20 +76,26 @@ func findLastKeyInAccountChangeSet(b []byte, k []byte, keyLen int) ([]byte, erro
 		return nil, fmt.Errorf("decode: input too short (%d bytes, expected at least %d bytes)", len(b), valOffset+totalValLength)
 	}
 
-	for i := n - 1; int(i) >= 0; i-- {
-		key := b[4+i*keyLen : 4+(i+1)*keyLen]
-		idx0 := 0
-		if i > 0 {
-			idx0 = int(binary.BigEndian.Uint32(b[4+n*keyLen+4*(i-1):]))
-		}
-		idx1 := int(binary.BigEndian.Uint32(b[4+n*keyLen+4*i:]))
-		val := b[valOffset+idx0 : valOffset+idx1]
+	id := sort.Search(n, func(i int) bool {
+		res := bytes.Compare(b[4+i*keyLen:4+(i+1)*keyLen], k)
+		return res >= 0
+	})
 
-		if bytes.Equal(key, k) {
-			return val, nil
-		}
+	if id >= n {
+		return nil, ErrNotFound
 	}
-	return nil, ErrNotFound
+
+	if !bytes.Equal(b[4+id*keyLen:4+(id+1)*keyLen], k) {
+		return nil, ErrNotFound
+	}
+
+	idx0 := 0
+	if id > 0 {
+		idx0 = int(binary.BigEndian.Uint32(b[4+n*keyLen+4*(id-1):]))
+	}
+
+	idx1 := int(binary.BigEndian.Uint32(b[4+n*keyLen+4*id:]))
+	return common.CopyBytes(b[valOffset+idx0 : valOffset+idx1]), nil
```

Comment:
Please do not copy by default here, the caller should be able to make the decision whether to copy or not

Reply:
Removed